File: package/go/tests/count_sales_test.go
package tests

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"program-knock/src"
)

func TestCountSales(t *testing.T) {
	t.Run("Basic", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Tokyo", PaymentMethod: "Credit", Product: "Apple", Quantity: 3},
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple", Quantity: 2},
			{Store: "Osaka", PaymentMethod: "Credit", Product: "Apple", Quantity: 5},
			{Store: "Tokyo", PaymentMethod: "Credit", Product: "Apple", Quantity: 1},
			{Store: "Osaka", PaymentMethod: "Credit", Product: "Orange", Quantity: 4},
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Banana", Quantity: 2},
			{Store: "Tokyo", PaymentMethod: "Credit", Product: "Apple", Quantity: 2},
		}

		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Tokyo", PaymentMethod: "Credit", Product: "Apple"}:  6, // 3 + 1 + 2
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple"}:    2, // 2
			{Store: "Osaka", PaymentMethod: "Credit", Product: "Apple"}:  5, // 5
			{Store: "Osaka", PaymentMethod: "Credit", Product: "Orange"}: 4, // 4
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Banana"}:   2, // 2
		}
		assert.Equal(t, expected, result)
	})

	t.Run("Empty", func(t *testing.T) {
		sales := []src.Sale{}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{}
		assert.Equal(t, expected, result)
	})

	t.Run("MultiplePaymentMethods", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Tokyo", PaymentMethod: "Credit", Product: "Apple", Quantity: 4},
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple", Quantity: 5},
			{Store: "Tokyo", PaymentMethod: "Credit", Product: "Apple", Quantity: 3},
			{Store: "Tokyo", PaymentMethod: "MobilePay", Product: "Apple", Quantity: 2},
		}

		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Tokyo", PaymentMethod: "Credit", Product: "Apple"}:    7, // 4 + 3
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple"}:      5, // 5
			{Store: "Tokyo", PaymentMethod: "MobilePay", Product: "Apple"}: 2, // 2
		}
		assert.Equal(t, expected, result)
	})

	t.Run("SingleSale", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Nagoya", PaymentMethod: "Cash", Product: "Grape", Quantity: 5},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Nagoya", PaymentMethod: "Cash", Product: "Grape"}: 5,
		}
		assert.Equal(t, expected, result)
	})

	t.Run("SameStoreDifferentProducts", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Fukuoka", PaymentMethod: "Credit", Product: "Melon", Quantity: 2},
			{Store: "Fukuoka", PaymentMethod: "Credit", Product: "Banana", Quantity: 3},
			{Store: "Fukuoka", PaymentMethod: "Credit", Product: "Melon", Quantity: 1},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Fukuoka", PaymentMethod: "Credit", Product: "Melon"}:  3, // 2 + 1
			{Store: "Fukuoka", PaymentMethod: "Credit", Product: "Banana"}: 3, // 3
		}
		assert.Equal(t, expected, result)
	})

	t.Run("QuantityMin", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple", Quantity: 1},
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple", Quantity: 1},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple"}: 2, // 1 + 1
		}
		assert.Equal(t, expected, result)
	})

	t.Run("LargeQuantity", func(t *testing.T) {
		largeNum := 1000000
		sales := []src.Sale{
			{Store: "Osaka", PaymentMethod: "Credit", Product: "Laptop", Quantity: largeNum},
			{Store: "Osaka", PaymentMethod: "Credit", Product: "Laptop", Quantity: largeNum},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Osaka", PaymentMethod: "Credit", Product: "Laptop"}: 2 * largeNum,
		}
		assert.Equal(t, expected, result)
	})

	t.Run("DuplicateSales", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Sapporo", PaymentMethod: "Credit", Product: "IceCream", Quantity: 3},
			{Store: "Sapporo", PaymentMethod: "Credit", Product: "IceCream", Quantity: 3},
			{Store: "Sapporo", PaymentMethod: "Credit", Product: "IceCream", Quantity: 3},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Sapporo", PaymentMethod: "Credit", Product: "IceCream"}: 9, // 3 + 3 + 3
		}
		assert.Equal(t, expected, result)
	})

	t.Run("CaseSensitivity", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "apple", Quantity: 2},
			{Store: "tokyo", PaymentMethod: "Cash", Product: "apple", Quantity: 4},
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple", Quantity: 1},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "apple"}: 2,
			{Store: "tokyo", PaymentMethod: "Cash", Product: "apple"}: 4,
			{Store: "Tokyo", PaymentMethod: "Cash", Product: "Apple"}: 1,
		}
		assert.Equal(t, expected, result)
	})

	t.Run("SpecialCharacters", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "", PaymentMethod: "PayPal", Product: "???", Quantity: 2},
			{Store: "", PaymentMethod: "PayPal", Product: "???", Quantity: 1},
			{Store: "New-Line\nStore", PaymentMethod: "Ca$h", Product: "Product\tTab", Quantity: 3},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "", PaymentMethod: "PayPal", Product: "???"}:                           3, // 2 + 1
			{Store: "New-Line\nStore", PaymentMethod: "Ca$h", Product: "Product\tTab"}: 3, // 3
		}
		assert.Equal(t, expected, result)
	})

	t.Run("ManyDifferentStores", func(t *testing.T) {
		sales := []src.Sale{
			{Store: "Store1", PaymentMethod: "Credit", Product: "Product1", Quantity: 1},
			{Store: "Store2", PaymentMethod: "Credit", Product: "Product1", Quantity: 2},
			{Store: "Store3", PaymentMethod: "Credit", Product: "Product1", Quantity: 3},
			{Store: "Store4", PaymentMethod: "Credit", Product: "Product1", Quantity: 4},
			{Store: "Store5", PaymentMethod: "Credit", Product: "Product1", Quantity: 5},
		}
		result := src.CountSales(sales)
		expected := map[src.SalesKey]int{
			{Store: "Store1", PaymentMethod: "Credit", Product: "Product1"}: 1,
			{Store: "Store2", PaymentMethod: "Credit", Product: "Product1"}: 2,
			{Store: "Store3", PaymentMethod: "Credit", Product: "Product1"}: 3,
			{Store: "Store4", PaymentMethod: "Credit", Product: "Product1"}: 4,
			{Store: "Store5", PaymentMethod: "Credit", Product: "Product1"}: 5,
		}
		assert.Equal(t, expected, result)
	})
}

File: package/typescript/tests/countSales.test.ts
import { countSales, Sale } from '../src/countSales';

describe('countSales', () => {
  // サンプル1
  test('basic test', () => {
    const sales: Sale[] = [
      { store: "Tokyo", paymentMethod: "Credit", product: "Apple", quantity: 3 },
      { store: "Tokyo", paymentMethod: "Cash", product: "Apple", quantity: 2 },
      { store: "Osaka", paymentMethod: "Credit", product: "Apple", quantity: 5 },
      { store: "Tokyo", paymentMethod: "Credit", product: "Apple", quantity: 1 },
      { store: "Osaka", paymentMethod: "Credit", product: "Orange", quantity: 4 },
      { store: "Tokyo", paymentMethod: "Cash", product: "Banana", quantity: 2 },
      { store: "Tokyo", paymentMethod: "Credit", product: "Apple", quantity: 2 },
    ];

    const result = countSales(sales);
    expect(result.get("Tokyo,Credit,Apple")).toBe(6); // 3 + 1 + 2
    expect(result.get("Tokyo,Cash,Apple")).toBe(2); // 2
    expect(result.get("Osaka,Credit,Apple")).toBe(5); // 5
    expect(result.get("Osaka,Credit,Orange")).toBe(4); // 4
    expect(result.get("Tokyo,Cash,Banana")).toBe(2); // 2
    expect(result.size).toBe(5);
  });

  // サンプル2
  test('empty sales', () => {
    const sales: Sale[] = [];
    const result = countSales(sales);
    expect(result.size).toBe(0);
  });

  // サンプル3
  test('multiple payment methods', () => {
    const sales: Sale[] = [
      { store: "Tokyo", paymentMethod: "Credit", product: "Apple", quantity: 4 },
      { store: "Tokyo", paymentMethod: "Cash", product: "Apple", quantity: 5 },
      { store: "Tokyo", paymentMethod: "Credit", product: "Apple", quantity: 3 },
      { store: "Tokyo", paymentMethod: "MobilePay", product: "Apple", quantity: 2 },
    ];

    const result = countSales(sales);
    expect(result.get("Tokyo,Credit,Apple")).toBe(7); // 4 + 3
    expect(result.get("Tokyo,Cash,Apple")).toBe(5); // 5
    expect(result.get("Tokyo,MobilePay,Apple")).toBe(2); // 2
    expect(result.size).toBe(3);
  });

  // 追加テストケース1: 単一の売上データ
  test('single sale', () => {
    const sales: Sale[] = [
      { store: "Nagoya", paymentMethod: "Cash", product: "Grape", quantity: 5 },
    ];
    const result = countSales(sales);
    expect(result.get("Nagoya,Cash,Grape")).toBe(5);
    expect(result.size).toBe(1);
  });

  // 追加テストケース2: 同一店舗・同一支払い方法で異なる商品
  test('same store different products', () => {
    const sales: Sale[] = [
      { store: "Fukuoka", paymentMethod: "Credit", product: "Melon", quantity: 2 },
      { store: "Fukuoka", paymentMethod: "Credit", product: "Banana", quantity: 3 },
      { store: "Fukuoka", paymentMethod: "Credit", product: "Melon", quantity: 1 },
    ];
    // Melon 2 + 1 = 3
    // Banana 3
    const result = countSales(sales);
    expect(result.get("Fukuoka,Credit,Melon")).toBe(3);
    expect(result.get("Fukuoka,Credit,Banana")).toBe(3);
    expect(result.size).toBe(2);
  });

  // 追加テストケース3: 数量が最小の境界値(1)
  test('quantity min', () => {
    const sales: Sale[] = [
      { store: "Tokyo", paymentMethod: "Cash", product: "Apple", quantity: 1 },
      { store: "Tokyo", paymentMethod: "Cash", product: "Apple", quantity: 1 },
    ];
    // 1 + 1 = 2
    const result = countSales(sales);
    expect(result.get("Tokyo,Cash,Apple")).toBe(2);
    expect(result.size).toBe(1);
  });

  // 追加テストケース4: 数量が非常に大きい場合
  test('large quantity', () => {
    const largeNum = 10 ** 6;
    const sales: Sale[] = [
      { store: "Osaka", paymentMethod: "Credit", product: "Laptop", quantity: largeNum },
      { store: "Osaka", paymentMethod: "Credit", product: "Laptop", quantity: largeNum },
    ];
    // 10^6 + 10^6 = 2 * 10^6
    const result = countSales(sales);
    expect(result.get("Osaka,Credit,Laptop")).toBe(2 * largeNum);
    expect(result.size).toBe(1);
  });

  // 追加テストケース5: 完全に同一のデータが複数ある場合
  test('duplicate sales', () => {
    const sales: Sale[] = [
      { store: "Sapporo", paymentMethod: "Credit", product: "IceCream", quantity: 3 },
      { store: "Sapporo", paymentMethod: "Credit", product: "IceCream", quantity: 3 },
      { store: "Sapporo", paymentMethod: "Credit", product: "IceCream", quantity: 3 },
    ];
    // 3 + 3 + 3 = 9
    const result = countSales(sales);
    expect(result.get("Sapporo,Credit,IceCream")).toBe(9);
    expect(result.size).toBe(1);
  });

  // 追加テストケース6: 大文字・小文字が違う店舗名/商品名
  test('case sensitivity', () => {
    const sales: Sale[] = [
      { store: "Tokyo", paymentMethod: "Cash", product: "apple", quantity: 2 },
      { store: "tokyo", paymentMethod: "Cash", product: "apple", quantity: 4 },
      { store: "Tokyo", paymentMethod: "Cash", product: "Apple", quantity: 1 },
    ];
    // (Tokyo, Cash, apple) = 2
    // (tokyo, Cash, apple) = 4
    // (Tokyo, Cash, Apple) = 1
    // それぞれ別のキーとして扱われる
    const result = countSales(sales);
    expect(result.get("Tokyo,Cash,apple")).toBe(2);
    expect(result.get("tokyo,Cash,apple")).toBe(4);
    expect(result.get("Tokyo,Cash,Apple")).toBe(1);
    expect(result.size).toBe(3);
  });

  // 追加テストケース7: 特殊文字や空文字列が含まれる場合
  test('special characters', () => {
    const sales: Sale[] = [
      { store: "", paymentMethod: "PayPal", product: "???", quantity: 2 },
      { store: "", paymentMethod: "PayPal", product: "???", quantity: 1 },
      { store: "New-Line\nStore", paymentMethod: "Ca$h", product: "Product\tTab", quantity: 3 },
    ];
    // ("", "PayPal", "???") = 2 + 1 = 3
    // ("New-Line\nStore", "Ca$h", "Product\tTab") = 3
    const result = countSales(sales);
    expect(result.get(",PayPal,???")).toBe(3);
    expect(result.get("New-Line\nStore,Ca$h,Product\tTab")).toBe(3);
    expect(result.size).toBe(2);
  });
});
